### ISSUE TITLE
Consider setting PYTHONUNBUFFERED.

### ISSUE THREAD

**Issue by GrahamDumpleton** [sclorg/s2i-python-container#157]
If a Python web application uses ``stdout`` for generating log messages, then the result can be buffered in memory and not output promptly such that the messages can be captured by Docker and OpenShift. The consequence of this is that there is potential for loss of log messages if the application crashes for some reason making it much harder to work out what the application was doing. The messages may also simply be delayed and not come out straight away.

It is not uncommon for people to set up the Python logging module to use ``stdout`` rather than ``stderr`` which exacerbates this problem.

Consideration should be given to setting the ``PYTHONUNBUFFERED`` environment variable as a default.

* https://docs.python.org/2/using/cmdline.html#envvar-PYTHONUNBUFFERED
* https://docs.python.org/2/using/cmdline.html#cmdoption-u

FWIW, Heroku sets ``PYTHONUNBUFFERED`` as a default as they likely have determined it is too confusing for users when they don't see their logged messages.

Note that this issue mainly affects gunicorn and hand rolled applications provided by ``app.py``. When using ``mod_wsgi-express`` it treats ``stdout`` the same as ``stderr`` and ensures that messages are always flushed when an EOL is encountered, with anything still remaining buffered at the end of a request which wasn't EOL terminated, being flushed automatically at the end of a request.

**Comment by GrahamDumpleton** [comment]
Note that it would be possible to undo this as the default by a user setting in the ``.s2i/environment`` file:

```
PYTHONUNBUFFERED=
```

That is, in the ``Dockerfile`` use:

```
ENV PYTHONUNBUFFERED=1
```

A user can undo it by setting the same environment variable, but to an empty string, thus overriding the default. This works as the Python interpreter code is:

```
    if (!saw_unbuffered_flag &&
        (p = Py_GETENV("PYTHONUNBUFFERED")) && *p != '\0')
        Py_UnbufferedStdioFlag = 1;
```

so only enables unbuffered ``stdio`` if the environment variable is non empty.

**Comment by ncoghlan** [comment]
For https://github.com/ncoghlan/srpminfo I'm seeing the problem where no logging output from the Python workers appears in the OpenShift console window, but setting `PYTHONUNBUFFERED=1` in the S2I environment didn't appear to make any difference.

**Comment by GrahamDumpleton** [comment]
Try using ``oc logs`` on pod. I have sometimes seen web console not showing logging. In your case the problem may be using gunicorn. It screws around with the stdout/stderr and can stuff things up. Maybe the case that default gunicorn options are suppressing it.

**Comment by GrahamDumpleton** [comment]
Possibly could be because you are using Flask.

* http://stackoverflow.com/a/30288342/128141

**Comment by ncoghlan** [comment]
I checked with https://github.com/ncoghlan/srpminfo/commit/82fa7cce1799e80d1e1520a5fa6d6a6047a3e4a6 (using explicit `print` calls to bypass the logging subsystem entirely) and it looks like the S2I image is configured to unconditionally throw away all console output from the WSGI process by default.

**Comment by ncoghlan** [comment]
Oops, @GrahamDumpleton pointed out that the code inside the `__main__` guard will never be called under gunicorn, so all my testing so far has been entirely invalid. Sorry for the noise :(

**Comment by rhcarvalho** [comment]
@GrahamDumpleton @sclorg/container-images this issue seems to be resolved already by #169.

**Comment by torsava** [comment]
Indeed, `PYTHONUNBUFFERED` is already being set thanks to #169. Closing the issue.
